homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,282 @@
+    {
+        vol.Required("issuer"): str,
+        vol.Required("jwks_uri"): str,
+        vol.Required("id_token_signing_alg_values_supported"): list,
+        vol.Optional("scopes_supported"): vol.Contains("openid"),
+        vol.Required("token_endpoint"): str,
+        vol.Required("authorization_endpoint"): str,
+        vol.Required("response_types_supported"): vol.Contains("code"),
+        vol.Optional(
+            "token_endpoint_auth_methods_supported", default=["client_secret_basic"]
+        ): vol.Contains("client_secret_post"),
+        vol.Optional(
+            "grant_types_supported", default=["authorization_code", "implicit"]
+        ): vol.Contains("authorization_code"),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when submitting invalid authentication."""
+
+
+async def raise_for_status(response: ClientResponse) -> None:
+    """Raise exception on data failure with logging."""
+    if response.status >= 400:
+        standard = ClientResponseError(
+            response.request_info,
+            response.history,
+            code=response.status,
+            headers=response.headers,
+        )
+        data = await response.text()
+        _LOGGER.error("Request failed: %s", data)
+        raise InvalidAuthError(data) from standard
+
+
+WANTED_SCOPES = {"openid", "email", "profile"}
+
+
+@AUTH_PROVIDERS.register("openid")
+class OpenIdAuthProvider(AuthProvider):
+    """Example auth provider based on hardcoded usernames and passwords."""
+
+    DEFAULT_TITLE = "OpenID Connect"
+
+    _configuration: Dict[str, Any]
+    _jwks: Dict[str, Any]
+
+    async def async_get_configuration(self) -> Dict[str, Any]:
+        """Get discovery document for OpenID."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self.config[CONF_CONFIGURATION]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], OPENID_CONFIGURATION_SCHEMA(data))
+
+    async def async_get_jwks(self) -> Dict[str, Any]:
+        """Get the keys for id verification."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self._configuration["jwks_uri"]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], data)
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        url = URL(get_url(self.hass, prefer_external=True, allow_cloud=False))
+        return str(url.with_path(AUTH_CALLBACK_PATH))
+
+    async def async_login_flow(self, context: Optional[Dict]) -> LoginFlow:
+        """Return a flow to login."""
+
+        if not hasattr(self, "_configuration"):
+            self._configuration = await self.async_get_configuration()
+
+        if not hasattr(self, "_jwks"):
+            self._jwks = await self.async_get_jwks()
```

Comment:
Should jwks expire and be re-fetched?

Reply:
Yes probably. Should look at the http headers and respect that.

Reply:
Do you know if we have some caching aiohttp client or if there is some lib? There are quite a few things to take into account to behave properly.